src/libs/elektra/hooks.c

```diff
@@ -0,0 +1,95 @@
+/**
+ * @file
+ *
+ * @brief
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include <kdbinternal.h>
+
+void freeHooks (KDB * kdb, Key * errorKey)
+{
+	if (kdb->hooks == NULL)
+	{
+		return;
+	}
+
+	if (kdb->hooks->gopts != NULL)
+	{
+		elektraPluginClose (kdb->hooks->gopts->plugin, errorKey);
+		free (kdb->hooks->gopts);
+		kdb->hooks->gopts = NULL;
+	}
+
+	free (kdb->hooks);
+	kdb->hooks = NULL;
+}
+
+static struct _HookPluginGopts * initHooksGopts (Plugin * plugin)
+{
+	if (!plugin)
+	{
+		return NULL;
+	}
+
+	struct _HookPluginGopts * hook = elektraCalloc (sizeof (struct _HookPluginGopts));
+	hook->plugin = plugin;
+
+	hook->kdbHookGoptsGet = (kdbHookGoptsGetPtr) elektraPluginGetFunction (plugin, "hooks/gopts/get");
+	return hook;
+}
+
+static Plugin * loadPlugin (const char * pluginName, KeySet * config, KeySet * modules, Key * errorKey)
+{
+	Key * openKey = keyDup (errorKey, KEY_CP_ALL);
+
+	Plugin * plugin = elektraPluginOpen (pluginName, modules, config, openKey);
+
+	if (!plugin)
+	{
+		ELEKTRA_ADD_INSTALLATION_WARNINGF (errorKey, "Could not load plugin '%s'", pluginName);
+		keyCopyAllMeta (errorKey, openKey);
+		keyDel (openKey);
+		return NULL;
+	}
+
+	return plugin;
+}
+
+static bool isGoptsEnabledByContract (KDB * kdb)
+{
+	bool goptsEnabled = false;
+
+	KeySet * ksTemp = ksDup (kdb->global);
+	Key * cutPoint = keyNew ("system:/elektra/gopts", KEY_END);
+
+	KeySet * cut = ksCut (ksTemp, cutPoint);
+
+	goptsEnabled = ksGetSize (cut) > 0;
+
+	keyDel (cutPoint);
+	ksDel (ksTemp);
+	ksDel (cut);
+
+	return goptsEnabled;
+}
```

Comment:
This is a bit of a hack. What you should really do is pass the `KeySet * contract` from `kdbOpen` to this function and just check for some key that is added by `elektraGOptsContract`.



You could just check whether `system:/elektra/contract/mountglobal/gopts` is there:

```suggestion
static bool isGoptsEnabledByContract (KeySet * contract)
{
	return ksLookupByName (contract, "system:/elektra/contract/mountglobal/gopts", 0) != NULL;
}
```

The `elektraGOptsContract` will always add this key and the [docs for the contract](https://www.libelektra.org/devdocu/kdb-contracts#mounting-global-plugins) also explain that.

In the long it's probably better (not least because of the `mountglobal` name) to also modify `elektraGOptsContract`. In the new hooks system the generic "mount plugin X with config Y" API doesn't really make sense. However, I don't think we should do this now. Instead I'd wait until we have also updated `spec`, `internalnotification`, etc. and can fully remove the global plugin system.